pkg/manifests/tls.go

```diff
@@ -0,0 +1,141 @@
+// without refresh time handling. We just take care of rotation if we reach 1/5 of the validity timespan before expiration.
+func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
+	maxWait := notAfter.Sub(notBefore) / 5
+	latestTime := notAfter.Add(-maxWait)
+	return now().After(latestTime)
+}
+
+func getSelfSignedCA(s *v1.Secret) (*crypto.CA, error) {
+	crt, crtPresent := s.Data["ca.crt"]
+	key, keyPresent := s.Data["ca.key"]
+
+	if crtPresent && keyPresent {
+		ca, err := crypto.GetCAFromBytes(crt, key)
+		if err != nil {
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
+		} else if !needsNewCert(ca.Config.Certs[0].NotBefore, ca.Config.Certs[0].NotAfter, time.Now) {
+			return nil, nil
+		}
+	}
+
+	cfg, err := crypto.MakeSelfSignedCAConfig(
+		fmt.Sprintf("%s@%d", "openshift-cluster-monitoring", time.Now().Unix()),
+		crypto.DefaultCertificateLifetimeInDays,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	crt, key, err = cfg.GetPEMBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	s.Data["ca.crt"] = crt
+	s.Data["ca.key"] = key
+
+	return &crypto.CA{
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+		Config:          cfg,
+	}, nil
+}
+
+// This method creates a central secret containing GRPC TLS key material
+// if the given secret is nil
+// or updates it if the CA present in the given secret is about to expire.
+//
+// It "rotates" the CA 1/5 before the expiry timespan.
+// The rotation scheme is very naive, it simply creates a new self signed CA
+// and refreshes all the subsequent GRPC server/client certs.
+// For simplicity, the CA as well as all certificates have the same expiration,
+// that is crypto.DefaultCertificateLifetimeInDays = 2 years
+// in order to align rotation.
+func (f *Factory) GRPCSecret(s *v1.Secret) (*v1.Secret, error) {
+	if s == nil {
+		var err error
+		s, err = f.NewSecret(MustAssetReader(ClusterMonitoringGrpcTLSSecret))
+		if err != nil {
+			return nil, err
+		}
+		s.Namespace = f.namespace
+		s.Data = make(map[string][]byte)
+		s.Annotations = make(map[string]string)
+	}
+
+	ca, err := getSelfSignedCA(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "error rotating CA")
+	}
+
+	// no need to rotate
+	if ca == nil {
+		return s, nil
+	}
+
+	{
+		cfg, err := ca.MakeClientCertificateForDuration(
+			&user.DefaultInfo{
+				Name: "thanos-querier",
+			},
+			time.Duration(crypto.DefaultCertificateLifetimeInDays)*24*time.Hour,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "error making thanos querier client certificate")
+		}
+		crt, key, err := cfg.GetPEMBytes()
+		if err != nil {
+			return nil, errors.Wrap(err, "error getting PEM bytes for thanos querier client certificate")
+		}
+		s.Data["thanos-querier-client.crt"] = crt
+		s.Data["thanos-querier-client.key"] = key
+	}
+
+	{
+		cfg, err := ca.MakeServerCert(
+			sets.NewString("prometheus-grpc"),
+			crypto.DefaultCertificateLifetimeInDays,
```

Comment:
This one doesn't need the `*24*time.Hour` like the other method?

Reply:
No, in that case the library expects an `int`: https://github.com/openshift/cluster-monitoring-operator/blob/c6195af44d80d4358f15f8e015c6cc608326f224/vendor/github.com/openshift/library-go/pkg/crypto/crypto.go#L665:15

Reply:
Ack